### PR TITLE
Implement new proc macro diagnostics API

### PR DESCRIPTION
**Pull request by jhpratt**
See [#54140 (comment)](https://github.com/rust-lang/rust/issues/54140#issuecomment-802701867) for details.

In this PR, I remove the existing `MultiSpan` trait, replacing it with the proposed `Spanned` trait. The `Diagnostic` API has been rebuilt from the base struct.

The `Level` enum still needs to be removed, as it's not used anywhere in the public API.

### CODE REVIEW

**Comment by jhpratt** [comment]
Not sure what needs to happen for this to be successful, as the changes are present in the `library/proc_macro` crate. @jyn514 suggested trying `#[cfg(bootstrap)]` on the change in the `compiler/rust_macros`, which I attempted to no avail. If someone could give me a nudge in the right direction, that would be great.

@rustbot modify labels to -S-waiting-on-author +s-waiting-on-review

File: library/proc_macro/src/diagnostic.rs
Comment (pickfire) [inline library/proc_macro/src/diagnostic.rs:185]: Do we need `Extend`?

File: compiler/rustc_expand/src/proc_macro_server.rs
Comment (pickfire) [inline compiler/rustc_expand/src/proc_macro_server.rs:640]: What happens if user `push_primary_span` multiple times? Does it become a secondary span?

**Comment by pickfire** [comment]
> Not sure what needs to happen for this to be successful, as the changes are present in the library/proc_macro crate. @jyn514 suggested trying #[cfg(bootstrap)] on the change in the compiler/rust_macros, which I attempted to no avail. If someone could give me a nudge in the right direction, that would be great.

```
#[cfg(bootstrap)]
use syn::spanned::Spanned;
```

I believe he meant this.

File: compiler/rustc_expand/src/proc_macro_server.rs
Comment (jhpratt) [reply compiler/rustc_expand/src/proc_macro_server.rs:640]: From what I can tell, "secondary spans" are those that are labelled. The method itself is identical, all I did was rename it to match the internal naming of "primary span" [[ref](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.MultiSpan.html)].

File: library/proc_macro/src/diagnostic.rs
Comment (jhpratt) [reply library/proc_macro/src/diagnostic.rs:185]: I don't see why this wouldn't work, so I'll give it a shot.

**Comment by jhpratt** [comment]
Ah, I misunderstood the statement about the `cfg` — I tried to gate what I'd already changed in that file behind the bootstrap.

Either way, though, I get the same error :disappointed: This is the case with either change or both changes.

**Comment by jhpratt** [comment]
Alright, I'm officially out of ideas. I am fairly certain I have the bootstrap cfgs backwards in the latest commit, but running `./x.py check --stage 1` locally succeeded; obviously it hasn't in CI. If I reverse the cfgs to what I believe is logical, it fails locally.

At this point I'll await input from others, as I don't particularly know what I'm doing here.

File: compiler/rustc_expand/src/proc_macro_server.rs
Comment (pickfire) [reply compiler/rustc_expand/src/proc_macro_server.rs:640]: > They can be primary spans. In this case they are the locus of the error, and would be rendered with ^^^.

Oh, there is just put a `push_primary_span` so I didn't know the context, the primary span that was added there is different than what I was thinking, I think we should either rename that or we should have some text letting others know what is primary span.

File: compiler/rustc_macros/src/session_diagnostic.rs
Comment (pickfire) [inline compiler/rustc_macros/src/session_diagnostic.rs:116]:
```suggestion
#[cfg(bootstrap)]
fn span_err(span: impl proc_macro::MultiSpan, msg: &str) -> proc_macro::Diagnostic {
    Diagnostic::spanned(span, proc_macro::Level::Error, msg)
}

#[cfg(not(bootstrap))]
fn span_err(span: impl proc_macro::Spanned, msg: &str) -> proc_macro::Diagnostic {
    Diagnostic::error(msg).mark_all(span)
}
```

Weird, seemed reversed to me, the errors seemed to show this part as well. Can you please try this?

Comment (jhpratt) [reply compiler/rustc_macros/src/session_diagnostic.rs:116]: That's what I tried at first. It didn't succeed locally.

Comment (pickfire) [reply compiler/rustc_macros/src/session_diagnostic.rs:116]: Not sure, maybe need someone else to help. cc @jyn514 to see if he can help.

Comment (petrochenkov) [reply compiler/rustc_macros/src/session_diagnostic.rs:116]: `rustc_macros` is a proc macro crate, cargo may pass rustc flags differently to it, so I'm not even sure whether `bootstrap` is passed to it at all.
If you don't want to debug the build issues, keeping the old API around will work around this problem for sure.

**Comment by petrochenkov** [comment]
This needs an API review from T-libs before a compiler implementation review.
r? @dtolnay 

Also cc @matklad regarding incremental proc macros.
Some span-related methods are not stabilized because they allow proc macros to inspect absolute positions in files, maybe the diagnostic API has some similar hazards.

File: compiler/rustc_macros/src/session_diagnostic.rs
Comment (jhpratt) [reply compiler/rustc_macros/src/session_diagnostic.rs:116]: Can't keep the entirety of the old API around, as some methods (such as `Diagnostic::error`) have the same name but change signatures. I know this because I've already tried :laughing:

File: compiler/rustc_expand/src/proc_macro_server.rs
Comment (jhpratt) [reply compiler/rustc_expand/src/proc_macro_server.rs:640]: Keep in mind that this detail is not exposed publicly. A user won't have to think about what kind of span is being added, as it's all in the same `mark`/`mark_all` API.

File: compiler/rustc_macros/src/session_diagnostic.rs
Comment (jyn514) [reply compiler/rustc_macros/src/session_diagnostic.rs:116]: I opened https://github.com/rust-lang/rust/pull/83433 which should fix this.

**Comment by jhpratt** [comment]
Expecting a failure from the latest push, just making sure I didn't _somehow_ do something stupid rebasing.

**Comment by jyn514** [comment]
> It was mentioned that this is only used by RLS (allowed breakage) and merge — he wasn't sure on whether this was allowed breakage or not.

To clarify: `merge` is used by bootstrap, which *can't* be broken, it's only built by the beta compiler. However this will block updating the bootstrap compiler. Fortunately this is the very beginning of the release cycle, so proc-macro-error has 6 weeks to adapt to the new API before it becomes a problem.

File: library/proc_macro/src/diagnostic.rs
Comment (pickfire) [inline library/proc_macro/src/diagnostic.rs:119]: Is this known as passthrough or delegate?

Comment (jhpratt) [reply library/proc_macro/src/diagnostic.rs:119]: To me delegation is this in reverse. Doesn't much matter to me either way, though.

File: library/proc_macro/src/diagnostic.rs
Comment (lambda-fairy) [inline library/proc_macro/src/diagnostic.rs:163]: Have we considered putting the `#[must_use]` on the `struct` definition instead?

I think that makes more sense than putting it on individual methods.

Comment (jhpratt) [reply library/proc_macro/src/diagnostic.rs:163]: Didn't know that was a thing. I'll change it once I'm on my laptop.

**Comment by crlf0710** [comment]
@jhpratt Ping from triage! CI is red here. mind fixing that?

**Comment by jyn514** [comment]
@crlf0710 this can't be fixed currently, it needs a way to pass cfg(bootstrap) to rustc_macros. https://github.com/rust-lang/rust/pull/83363#issuecomment-806194793

**Comment by jhpratt** [comment]
Once that is fixed, though, it should be trivial to get CI green!

**Comment by jhpratt** [comment]
Happy to address merge conflicts when the upstream bug fixes are complete.

**Comment by jyn514** [comment]
@jhpratt I won't have time to work on https://github.com/rust-lang/rust/pull/83433 for the foreseeable future.

**Comment by jhpratt** [comment]
@jyn514 What should I do with this PR in that situation?

**Comment by jyn514** [comment]
I don't know, sorry. I don't have time to look into this.

**Comment by jyn514** [comment]
I'm happy for you to work on https://github.com/rust-lang/rust/pull/83433 instead of me if you're interested, but I don't have time to offer any guidance.

**Comment by jhpratt** [comment]
Not an issue — I just wasn't sure if I should leave this PR open for the most part. I don't have the requisite knowledge (or time to attain it) for #83433. Not a big deal holding off 🙂

**Comment by jhpratt** [comment]
@rustbot modify labels to: -S-blocked +S-waiting-on-author

Now that the blocker has been resolved, I'm going to continue work on this. The PR appears to still be in good shape. What I am currently working on is making it so the `proc_macro::Level` enum is no longer public, as it's not used anywhere.

File: src/test/ui/proc-macro/auxiliary/custom-quote.rs
Comment (jhpratt) [inline src/test/ui/proc-macro/auxiliary/custom-quote.rs:10]: To answer the inevitable question, this is necessary because the new `MultiSpan` trait has a `.span()` method, which would override the inherent method when both are in scope.

**Comment by jhpratt** [comment]
CI is passing and code is in decent shape. If anyone could point me in the right direction for how to make `proc_macro::Level` non-public, that would be great; right now it's just interaction with the proc macro bridge that's the issue. However, I don't see that as a reason to block this PR from being merged, seeing as this is all unstable to begin with. I have doc-hidden the `Level` enum and moved it to the `proc_macro_internals` feature, such that it's not exposed under ordinary circumstances.

cc @dtolnay for review

**Comment by jhpratt** [comment]
Also cc @SergioBenitez in case you want to take a look.

**Comment by jhpratt** [comment]
Ping @dtolnay for review. Or anyone else for that matter. Happy to address merge conflicts before a review if desired.

File: compiler/rustc_macros/src/session_diagnostic.rs
Comment (lambda-fairy) [inline compiler/rustc_macros/src/session_diagnostic.rs:107]: Does this need to be a macro? This seems like it can be a normal function, like it was before.

File: library/proc_macro/src/diagnostic.rs
Comment (lambda-fairy) [inline library/proc_macro/src/diagnostic.rs:13]: I'm not a fan of keeping commented code around. On first read I thought this was a typo (or bad merge).

Can we replace this with only a TODO that links to the relevant issue?

Same with the `fn warning` definition below.

**Review by lambda-fairy — CHANGES_REQUESTED** [review]
I only have a couple nits to contribute, but I hope it helps regardless 😄

File: compiler/rustc_macros/src/session_diagnostic.rs
Comment (jhpratt) [reply compiler/rustc_macros/src/session_diagnostic.rs:107]: I can add a comment, but the macro is intended to be temporary until a bootstrap takes place, hence the implementation. The method name changed but the signature stayed the same, so it was easier to just do this.

File: library/proc_macro/src/diagnostic.rs
Comment (jhpratt) [reply library/proc_macro/src/diagnostic.rs:13]: Yeah, probably best to just remove it outright, given that the form lint-associated warnings take will certainly require a different API.

**Comment by jyn514** [comment]
r? rust-lang/libs

**Comment by jhpratt** [comment]
Still waiting on a review here from someone. As I said I can address the merge conflict before a review if desired.

**Comment by yaahc** [comment]
> Still waiting on a review here from someone.

Sorry about that :sweat_smile:. I can review this today or tomorrow but can you do me a favor and cleanup the comment history as much as possible first? I started reviewing and noticed some of the earlier comments have already been hidden, but it's unclear to me if the earlier un-hidden comments (for example, [[1]](https://github.com/rust-lang/rust/pull/83363#issuecomment-803761046), [[2]](https://github.com/rust-lang/rust/pull/83363#issuecomment-804223135)) are still relevant or not ([[3]](https://github.com/rust-lang/rust/pull/83363#issuecomment-806194793) seems to imply they've been resolved). If they're all relevant then lmk and I'll start the review, but otherwise I'll wait so I don't waste effort reading out of date info.

> As I said I can address the merge conflict before a review if desired.

I'm not as concerned about this part, feel free to leave the conflict or resolve it before I review it based on your own preference.

**Comment by yaahc** [comment]
I went ahead and marked 1 and 2 as resolved since I don't think either of those concerns are particularly important. Those can be resolved in follow up PRs if anyone feels a particular need. The third one about cleaning up the comment seems justified to me.

Also, the PR looks solid but I have some concerns with the proposed API so I went ahead and left a comment on the tracking issue rather than here.

**Comment by jhpratt** [comment]
Sounds good. I'll rebase and remove the commented out code.

Going to mark this blocked pending discussion in the tracking issue.

@rustbot label: -S-waiting-on-review +S-blocked

**Comment by JohnCSimon** [comment]
@jhpratt 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Comment by jhpratt** [comment]
_I_ am ready to continue with it. There was just concerns regarding the exact API, which is unfortunate. It's an issue for T-libs-api, ultimately.

Thank you for the reminder of GitHub's quirk. I was aware of it previously, but a reminder never hurts.

**Comment by JohnCSimon** [comment]
@jhpratt ok I'll reopen it. We have a lot of dead PRs.

**Comment by jhpratt** [comment]
Oh, I understand 100%. It just needs a decision from T-libs-api to my recollection. I am more than happy to rebase and have this ready to merge if it's desired.

cc @rust-lang/libs-api — any chance of giving this another look?

**Comment by estebank** [comment]
Hi, @rust-lang/libs! Would it be possible to get a review of this PR going? I would like to see us make progress in stabilizing (at least a part of) the `Diagnostic` API sometime before edition 2024 comes out.

**Comment by dtolnay** [comment]
I was supposed to look at this but syn 2-related work has been more important so far. Anyone else should feel free to weigh in with their feedback.

**Comment by jhpratt** [comment]
With regard to the implementation itself (given the team pinged), it's been so long that a rebase will be _significant_ effort certainly involving nontrivial changes. It's more the API that I'd like to see agreed upon first — the implementation would naturally follow.

Given the duration, it may make sense to redo the implementation from scratch. Not certain on that, though.

**Comment by m-ou-se** [comment]
Is there an accurate API overview for this change? That'd make it a lot easier to discuss the proposed API. :)

**Comment by m-ou-se** [comment]
It seems that this is the API introduced by this PR:

```rust
pub trait Spanned {
    type Iter: Iterator<Item = Span>;
    fn spans(self) -> Self::Iter;
    fn span(self) -> Span where Self: Sized;
    fn error(self, msg: &str) -> Diagnostic where Self: Sized;
    // no warning
    fn note(self, msg: &str) -> Diagnostic where Self: Sized;
}

impl<S: Spanned, I: IntoIterator<Item = S>> Spanned for I;
impl Spanned for Span;
impl Spanned for Group;
impl Spanned for Ident;
impl Spanned for Literal;
impl Spanned for Punct;
impl Spanned for TokenTree;

#[must_use]
struct Diagnostic { … }

impl Diagnostic {
    pub fn error(message: &str) -> Self;
    // no warning
    pub fn note(message: &str) -> Self;
    pub fn with_help(mut self, message: &str) -> Self;
    pub fn with_note(mut self, message: &str) -> Self;
    pub fn mark(mut self, item: impl Spanned) -> Self;
    pub fn mark_all(mut self, item: impl Spanned) -> Self;
    pub fn label(mut self, item: impl Spanned, message: &str) -> Self;
    pub fn emit(self);
}
```

In [this comment](https://github.com/rust-lang/rust/issues/54140#issuecomment-802701867), it was noted that:

- This API allows for diagnostics without a span, which might be fine.
- We might want to change `&str` to `Into<String>` or something similar.
- We probably want to add `&mut self` methods (`add_help`, `add_note`, etc.) too.
- This doesn't support suggestions yet.
- A `lint` mechanism would be useful for warnings, so they can be silenced with `#[allow()]`.

**Comment by m-ou-se** [comment]
`mark`/`mark_all` don't seem to have an equivalent on rustc's internal Diagnostic api. What does/should they do exactly? In the implementation in this PR, they seem to add spans to `MultiSpan::primary_spans`, but in the description in [this comment](https://github.com/rust-lang/rust/issues/54140#issuecomment-802701867) describes "secondary" spans, which sound more like the equivalent of an empty `.label(span, "")`?

**Comment by m-ou-se** [comment]
I found it useful to have a little overview of the current structure a [`Diagnostic`](https://doc.rust-lang.org/1.68.0/nightly-rustc/rustc_errors/struct.Diagnostic.html) in rustc itself. The current structure is roughly:

<pre>
Diagnostic {
    <a href="https://doc.rust-lang.org/1.68.0/nightly-rustc/rustc_errors/enum.Level.html">level</a>,
    message,
    code: Option&lt;<a href="https://doc.rust-lang.org/1.68.0/nightly-rustc/rustc_errors/diagnostic/enum.DiagnosticId.html">error or lint name</a>&gt;,
    primary_spans: [Span],
    span_labels: [(Span, String)],
    children: [
        SubDiagnostic {
            level,
            message,
            primary_spans: [Span],
            span_labels: [(Span, String)],
        }
    ],
    suggestions: [
        Suggestion {
            <a href="https://doc.rust-lang.org/1.68.0/nightly-rustc/rustc_errors/struct.CodeSuggestion.html#structfield.substitutions">substitutions</a>: [[(Span, String)]],
            message,
            <a href="https://doc.rust-lang.org/1.68.0/nightly-rustc/rustc_errors/enum.SuggestionStyle.html">style</a>,
            <a href="https://doc.rust-lang.org/1.68.0/nightly-rustc/rustc_errors/enum.Applicability.html">applicability</a>,
        }
    ],
}
</pre>

(The actual type also supports [localisation](https://doc.rust-lang.org/1.68.0/nightly-rustc/rustc_errors/enum.DiagnosticMessage.html) (with [placeholders](https://doc.rust-lang.org/1.68.0/nightly-rustc/rustc_errors/struct.Diagnostic.html#structfield.args)) for the strings, and [styling](https://doc.rust-lang.org/1.68.0/nightly-rustc/rustc_errors/snippet/enum.Style.html) for the messages.)

**Comment by m-ou-se** [comment]
Esteban and I came up with an alternative idea. See [this comment](https://github.com/rust-lang/rust/issues/54140#issuecomment-1487529341). Curious to hear what you think.

**Comment by Mark-Simulacrum** [comment]
It looks to me like discussion on https://github.com/rust-lang/rust/issues/54140 is/was ongoing since Mara's comment and there was positive reception to a different API shape than this PR currently implements. I'm going to go ahead and close as such. I think this can be reopened (or perhaps a new PR, to avoid running into github limitations on hiding comments to the extent possible) if there's a ready proposal for API shape that is positively received (not sure based on quick skim of the issue thread).

**Comment by jhpratt** [comment]
I'm not sure calling it "positive reception" is entirely fair, as there were significant questions raised, including some by myself that were never even remotely addressed.

Please don't take this personally, but this PR has been extremely frustrating. I opened it nearly three years ago, expecting it to be approved quite quickly as the general shape was agreed upon before I even began implementation. After multiple follow-ups, rebases, and trying to get the necessary reviews, it's just outright closed with what I consider a mischaracterization of people's feelings.

This is why people are hesitant to contribute to the standard library. It's not _that_ technically challenging. It's that it's damn near impossible to get anything nontrivial done unless you're on a team, and joining teams has no apparent path. Even when things _can_ get done, it typically takes at least a month to get even a preliminary review, with a similar timeframe for subsequent reviews. People _want_ to do things, but the process is extraordinarily unwelcoming at every step.

**Pull request closed without merging**
